### ISSUE TITLE
Should joins with different variables use rename?

### ISSUE THREAD

**Issue by hadley** [tidyverse/dtplyr#239]
``` r
library(dplyr, warn.conflicts = FALSE)
library(dtplyr)
library(nycflights13)

flights_dt <- lazy_dt(flights)
airports_dt <- lazy_dt(airports)

flights_dt %>%
  left_join(airports_dt, by = c("dest" = "faa")) %>%
  show_query()
#> merge(`_DT1`, `_DT2`, all.x = TRUE, all.y = FALSE, by.x = "dest", 
#>     by.y = "faa", allow.cartesian = TRUE)

flights_dt %>%
  left_join(airports_dt %>% rename(dest = faa)) %>%
  show_query()
#> Joining, by = "dest"
#> setnames(copy(`_DT2`), "faa", "dest")[`_DT1`, on = .(dest), allow.cartesian = TRUE]
```

<sup>Created on 2021-05-06 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

Not sure if this has any performance difference or not

**Comment by myoung3** [comment]
setnames has virtually no performance impact.  see matt's column on the related setcolorder here: https://stackoverflow.com/questions/12232041/how-to-reorder-data-table-columns-without-copying

**Comment by markfairbanks** [comment]
The way the query is currently constructed requires ``copy(`_DT2`)`` since the rename happens before the join. Then another "copy" occurs to create the new joined data.table.

If the query is switched over to using a named vector in `on`, using `copy()` can be avoided, and `setnames()` won't alter the original data.tables.

``` r
library(data.table)
library(nycflights13)
options(datatable.print.trunc.cols = TRUE)

flights_dt <- as.data.table(flights)
airports_dt <- as.data.table(airports)

result_dt <- airports_dt[flights_dt, on = c("faa" = "dest"), allow.cartesian = TRUE]

setnames(result_dt, "faa", "dest")

result_dt
#>         dest                            name      lat       lon  alt tz  dst
#>      1:  IAH    George Bush Intercontinental 29.98443 -95.34144   97 -6    A
#>      2:  IAH    George Bush Intercontinental 29.98443 -95.34144   97 -6    A
#>      3:  MIA                      Miami Intl 25.79325 -80.29056    8 -5    A
#>      4:  BQN                            <NA>       NA        NA   NA NA <NA>
#>      5:  ATL Hartsfield Jackson Atlanta Intl 33.63672 -84.42807 1026 -5    A
#>     ---                                                                     
#> 336772:  DCA   Ronald Reagan Washington Natl 38.85208 -77.03772   15 -5    A
#> 336773:  SYR           Syracuse Hancock Intl 43.11119 -76.10631  421 -5    A
#> 336774:  BNA                  Nashville Intl 36.12447 -86.67819  599 -6    A
#> 336775:  CLE          Cleveland Hopkins Intl 41.41169 -81.84979  791 -5    A
#> 336776:  RDU             Raleigh Durham Intl 35.87764 -78.78747  435 -5    A
#> 19 variables not shown: [tzone, year, month, day, dep_time, sched_dep_time, dep_delay, arr_time, sched_arr_time, arr_delay, ...]
```

**Comment by markfairbanks** [comment]
Closing as this was addressed by #240
